keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -161,273 +122,289 @@ consistent with the syntax supported by any of the common proxy vendors:
+
+TODO: We have a choice here: roundtrip with the annotation to class
+(e.g. take the annotation value and make it the `Class` string. Or
+support both values and prioritize the annotation as the class.
+
 ### Alternative backend types
 
-The Ingress resource is an L7 description of a composite set of services. It
-current supports Kubernetes Services as a backends. However, there are many use
-cases where a portion of the HTTP requests could be routed to a different kind
-of resource. For example, serving content from an object storage
-([S3][s3-backend], [GCS][gcs-backend]) is a commonly requested feature.
+The Ingress resource is an L7 description of a composite set of
+services. It current supports Kubernetes Services as a backends.
+However, there are many use cases where a portion of the HTTP requests
+could be routed to a different kind of resource. For example, serving
+content from an object storage ([S3][s3-backend], [GCS][gcs-backend])
+is a commonly requested feature.
 
-At the same time, we do not expect to enumerate all possible backends that could
-arise, nor do we expect that naming of the resources will be uniform in schema,
-parameters etc. Similarly, many of the resources will be platform specific.
+At the same time, we do not expect to enumerate all possible backends
+that could arise, nor do we expect that naming of the resources will
+be uniform in schema, parameters etc. Similarly, many of the resources
+will be platform specific.
 
-#### Proposal
+#### Backend types proposal
 
 Add a field to the `IngressBackend` struct with an object reference:
 
 ```golang
 type IngressBackend struct {
-  // Specifies the name of the referenced service. The service must exist in
-  // the same namespace as the Ingress object. The API server will
-  // automatically populate the Resource field below.
-  // +optional
-  ServiceName string
+  // Only one of the following fields may be specified.
 
-  // Specifies the port of the referenced service. If unspecfied and the
-  // ServiceName is non-empty, the Service must expose a single port.
+  // Service references a Service as a Backend. This is specially
+  // called out as it is required to be supported AND to reduce
+  // verbosity.
   // +optional
-  ServicePort intstr.IntOrString
+  Service *ServiceBackend
 
   // Resource is an ObjectRef to another Kubernetes resource in the namespace
-  // of the Ingress object. If Resource is a Service, then ServiceName will
-  // populated to match the Resource. If both ServiceName and Resource are
-  // specified, then they must reference the same Service object.
+  // of the Ingress object.
+  // +optional
+  Resource *v1.TypedLocalObjectReference
+}
+
+// ServiceBackend references a Kubernetes Service as a Backend.
+type ServiceBackend struct {
+  // Service is the name of the referenced service. The service must exist in
+  // the same namespace as the Ingress object.
+  // +optional
+  Service string
```

Comment:
shouldnt this be called "Name"

Reply:
This was in response to @apelisse 's comment about the use of the "Name" field as reserved for key-like.
It is more natural as "Name".